src/wallet/wallet.cpp

```diff
@@ -426,27 +442,243 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
+        return true;
+    }
+
+    return CWalletDB(strWalletFile).WriteUnifiedFullViewingKey(ufvk);
+}
+
+std::optional<ZcashdUnifiedFullViewingKey> CWallet::GetUnifiedFullViewingKeyByAccount(libzcash::AccountId accountId) {
+    if (!mnemonicHDChain.has_value()) {
+        throw std::runtime_error(
+                "CWallet::GetUnifiedFullViewingKeyByAccount(): Wallet is missing mnemonic seed metadata.");
+    }
+
+    auto seedfp = mnemonicHDChain.value().GetSeedFingerprint();
+    auto entry = mapUnifiedAccountKeys.find(std::make_pair(seedfp, accountId));
+    if (entry != mapUnifiedAccountKeys.end()) {
+        return CCryptoKeyStore::GetUnifiedFullViewingKey(entry->second);
     } else {
         return std::nullopt;
     }
 }
 
+UAGenerationResult CWallet::GenerateUnifiedAddress(
+    const libzcash::AccountId& accountId,
+    const libzcash::diversifier_index_t& j,
+    const std::set<libzcash::ReceiverType>& receiverTypes)
+{
+    if (!libzcash::HasShielded(receiverTypes)) {
+        return AddressGenerationError::InvalidReceiverTypes;
+    }
+
+    // The wallet must be unlocked in order to generate new transparent UA
+    // receivers, because we need to be able to add the secret key for the
+    // external child address at the diversifier index to the wallet's
+    // transparent backend in order to be able to detect transactions as
+    // ours rather than considering them as watch-only.
+    bool hasTransparent = receiverTypes.find(ReceiverType::P2PKH) != receiverTypes.end();
+    if (hasTransparent) {
+        if (!j.ToTransparentChildIndex().has_value()) {
+            return AddressGenerationError::InvalidTransparentChildIndex;
+        }
+
+        if (IsCrypted() || !GetMnemonicSeed().has_value()) {
+            return AddressGenerationError::WalletEncrypted;
+        }
+    }
+
+    auto ufvk = GetUnifiedFullViewingKeyByAccount(accountId);
+    if (ufvk.has_value()) {
+        auto ufvkid = ufvk.value().GetKeyID();
+
+        // Check whether an address has already been generated for this
+        // diversifier index. If so, ensure that the set of receiver types
+        // being requested is the same as the set of receiver types that was
+        // previously generated & return the previously generated address,
+        // otherwise return an error.
+        auto metadata = mapUfvkAddressMetadata.find(ufvkid);
+        if (metadata != mapUfvkAddressMetadata.end()) {
+            auto receivers = metadata->second.GetReceivers(j);
+            if (receivers.has_value()) {
+                if (receivers.value() == receiverTypes) {
+                    return std::make_pair(ufvk.value().Address(j, receiverTypes).value(), j);
+                } else {
+                    return AddressGenerationError::ExistingAddressMismatch;
+                }
+            }
+        }
+
+        // Find a working diversifier and construct the associated address.
+        auto foundAddress = ufvk.value().FindAddress(j, receiverTypes);
```

Comment:
The found diversifier index could be past the end of the transparent range.

Reply:
Fixed in f86d04f266308384be35e477b6b9e501d142b17f